src/vm/eventtrace.cpp

```diff
@@ -6241,6 +6243,79 @@ VOID ETW::LoaderLog::SendModuleEvent(Module *pModule, DWORD dwEventOptions, BOOL
     }
 }
 
+VOID ETW::MethodLog::SendMethodDetailsEvent(MethodDesc *pMethodDesc)
+{
+    CONTRACTL {
+        NOTHROW;
+        GC_NOTRIGGER;
+    } CONTRACTL_END;
+
+    EX_TRY
+    {
+        if(ETW_TRACING_CATEGORY_ENABLED(MICROSOFT_WINDOWS_DOTNETRUNTIME_PROVIDER_DOTNET_Context, 
+                                        TRACE_LEVEL_INFORMATION, 
+                                        CLR_METHODDIAGNOSTIC_KEYWORD))
+        {
+            if (pMethodDesc->IsDynamicMethod())
+                goto done;
+
+            Instantiation inst = pMethodDesc->GetMethodInstantiation();
+
+            if (inst.GetNumArgs() > 1024) // ETW has a limit for maximum event size. Do not log overly large method type argument sets
+                goto done;
+
+            BulkTypeEventLogger typeLogger;
+
+            ULONGLONG typeID = (ULONGLONG)pMethodDesc->GetMethodTable_NoLogging();
+            ETW::TypeSystemLog::LogTypeAndParametersIfNecessary(&typeLogger, typeID, ETW::TypeSystemLog::kTypeLogBehaviorTakeLockAndLogIfFirstTime);
+            ULONGLONG loaderModuleID = (ULONGLONG)pMethodDesc->GetLoaderModule();
+
+            // We're about to recursively call ourselves for the type parameters, so make a
+            // local copy of their type handles first (else, as we log them we could flush
+            // and clear out m_rgBulkTypeValues, thus trashing pVal)
+            
+            StackSArray<ULONGLONG> rgTypeParameters;
+            DWORD cParams = inst.GetNumArgs();
+
+            BOOL fSucceeded = FALSE;
+            EX_TRY
+            {
+                for (COUNT_T i = 0; i < cParams; i++)
+                {
+                    rgTypeParameters.Append((ULONGLONG)inst[i].AsPtr());
```

Comment:
Do the type parameters added here include the type parameters for generic classes that these methods may belong to? E.g. If I have `MyClass<A>.Foo<B>()` then do the type parameters consist of A and B or just B? If just B, is there a way to determine that MyClass has a type argument of A?

Reply:
The type parameters here are only the generic method's type parameters. The generic parameters of any containing type are held in the owningType output (and a consumer of this event is expected to decode the BulkType event to determine what those parameters might be. So, this event will only report B, but A may be acquired through the type event.